staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go

```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
 func (t *mockTyper) Recognizes(_ schema.GroupVersionKind) bool {
 	return false
 }
+
+func TestSerializerEncodeWithAllocator(t *testing.T) {
+	testCases := []struct {
+		name    string
+		gvk     schema.GroupVersionKind
+		creater runtime.ObjectCreater
+		obj     runtime.Object
+	}{
+		// scenario 1
+		{
+			name:    "roundtrip a bufferedMarshaller obj",
+			gvk:     schema.GroupVersionKind{Group: "group", Version: "version", Kind: "Carp"},
+			creater: &mockCreater{obj: &testapigroupv1.Carp{}},
+			obj: &testapigroupv1.Carp{
+				TypeMeta: metav1.TypeMeta{APIVersion: "group/version", Kind: "Carp"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "name",
+					Namespace: "namespace",
+				},
+				Spec: testapigroupv1.CarpSpec{
+					Subdomain: "carp.k8s.io",
+				},
+			},
+		},
+
+		// scenario 2
+		{
+			name:    "roundtrip a runtime.Unknown obj",
+			gvk:     schema.GroupVersionKind{Group: "group", Version: "version", Kind: "Unknown"},
+			creater: &mockCreater{obj: &runtime.Unknown{}},
+			obj:     &runtime.Unknown{TypeMeta: runtime.TypeMeta{APIVersion: "group/version", Kind: "Unknown"}},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// test data
+			writer := &bytes.Buffer{}
+			target := NewSerializer(tc.creater, &mockTyper{gvk: &tc.gvk})
+
+			// act
+			// step 1: perform enc/dec of the obj using target.Encode method
+			if err := target.Encode(tc.obj, writer); err != nil {
+				t.Fatal(err)
+			}
+			decodedObject, _, err := target.Decode(writer.Bytes(), &tc.gvk, nil)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !cmp.Equal(tc.obj, decodedObject) {
+				t.Fatalf("a test object encoded with the Encode method has unexpected data after deserializaiton, diff: %s", cmp.Diff(tc.obj, decodedObject))
+			}
+
+			// step 2: perform enc/dec of the obj using target.EncodeWithAllocator method
+			writer = &bytes.Buffer{}
+			alloc := &testAllocator{}
+			if err := target.EncodeWithAllocator(tc.obj, writer, alloc); err != nil {
+				t.Fatal(err)
+			}
+			if alloc.allocateCount != 1 {
+				t.Fatalf("expected the Allocate method to be called exactly 1 but it was executed: %v times ", alloc.allocateCount)
+			}
+			decodedObject, _, err = target.Decode(writer.Bytes(), &tc.gvk, nil)
```

Comment:
Instead of testing Decoding again here, I would actually ensure that the serialized version is exactly the same.

So let's maybe test if the writer1.Bytes() are equal to writer2.Bytes()

Reply:
are you afraid that the new encoding method would add some data and the `Decode` method would actually remove it?

Reply:
So conceptually what we need is to ensure bckward-compatibility of the wire format (which is effectively a serialized data).

Decoder may ignore some data (and in fact IIRC it already drops unknown extensions). So if the encoder would be adding some well-structured data, it's completely possible that decoder would actually drop it.